python/tskit/trees.py

```diff
@@ -833,6 +833,57 @@ def unrank(rank, num_leaves):
         """
         return combinatorics.RankTree.unrank(rank, num_leaves).to_tsk_tree()
 
+    def count_topologies(self, sample_sets=None):
+        """
+        Calculates the distribution of embedded topologies for every combination
+        of the sample sets in ``sample_sets``. ``sample_sets`` defaults to all
+        samples in the tree grouped by population. In this case, the topologies
+        produced represent evolutionary topologies of the populations in this
+        tree.
+
+        The returned object is a :class:`tskit.TopologyCounter` that contains
+        counts of topologies per combination of sample sets. For example,
+
+        >>> topology_counter = tree.count_topologies()
+        >>> rank, count = topology_counter[0, 1, 2].most_common(1)[0]
+
+        produces the most common tree topology, with populations 0, 1
+        and 2 as its tips, according to the genealogies of those
+        populations' samples in this tree.
+
+        The counts for each topology in the :class:`tskit.TopologyCounter`
+        are absolute counts that we would get if we were to select all
+        combinations of samples from the relevant sample sets.
+        For sample sets :math:`[s_0, s_1, ..., s_n]`, the total number of
+        topologies for those sample sets is equal to
+        :math:`|s_0| * |s_1| * ... * |s_n|`, so the counts in the counter
+        ``topology_counter[0, 1, ..., n]`` should sum to
+        :math:`|s_0| * |s_1| * ... * |s_n|`.
+
+        To convert the topology counts to probabilities, divide by the total
+        possible number of sample combinations from the sample sets in question::
+
+            >>> set_sizes = [len(sample_set) for sample_set in sample_sets]
+            >>> p = count / (set_sizes[0] * set_sizes[1] * set_sizes[2])
+
+        .. warning:: The interface for this method is preliminary and may be subject to
+            backwards incompatible changes in the near future.
+
+        :param list sample_sets: A list of lists of Node IDs, specifying the
+            groups of nodes to compute the statistic with.
+            Defaults to all samples grouped by population.
+        :rtype: tskit.TopologyCounter
+        :raises ValueError: If nodes in ``sample_sets`` are invalid or are
+            internal samples.
+        """
+        if sample_sets is None:
+            sample_sets = [
+                np.array(self.tree_sequence.samples(population=pop.id))
```

Comment:
We don't need this cast, do we? I'm sure tree_sequence.samples returns a numpy array.